src/substrait/expression/DecimalLiteral.cpp

```diff
@@ -0,0 +1,95 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include "substrait/expression/DecimalLiteral.h"
+
+#include <sstream>
+
+#include "absl/numeric/int128.h"
+#include "substrait/proto/algebra.pb.h"
+
+namespace io::substrait::expression {
+
+namespace {
+
+// invertBits flips the sign of a two-complements value.
+std::string invertBits(const std::string& value) {
+  std::string newValue = value;
+  // Flip all of the bits and add one.
+  bool carryover = true;
+  for (char& b : newValue) {
+    uint8_t newB = ~(static_cast<uint8_t>(b));
+    if (carryover) {
+      newB++;
+    }
+    carryover = carryover && (newB == 0);
+    b = static_cast<char>(newB & 0xff);
+  }
+  return newValue;
+}
+
+} // namespace
+
+DecimalLiteral DecimalLiteral::fromProto(
+    const ::substrait::proto::Expression_Literal_Decimal& proto) {
+  return {proto.value(), proto.precision(), proto.scale()};
+}
+
+bool DecimalLiteral::isValid() {
+  return value_.size() == 16 && precision_ >= 1 && precision_ <= 38;
```

Comment:
Similar to how we handle round, I believe scale can be negative to move the decimal place left of the decimal point (adding powers of 10).  Precision and scale don't appear to be related as you could be working with something with the scale of microseconds but only have 2 digits of precision.

Reply:
I see what you're saying but the spec does say 

![image](https://user-images.githubusercontent.com/1696093/229882200-36b0d00d-946a-489a-82f0-44554e2452f6.png)

and this is in line with other engines such as SQL server:

![image](https://user-images.githubusercontent.com/1696093/229882419-b4a71b89-373f-4b0f-85e4-2f7d981e4624.png)

I think the basic problem is that `P` is the "number of digits" and not "number of significant digits".  So you can't have decimal<2, -2> to represent something like `1100` (eleven hundred) because then you'd have 4 digits.

However, we can worry about this in follow-up / future PRs.  I'll open an issue.